tests/state_test_util.go

```diff
@@ -175,63 +230,76 @@ func (t *StateTest) Run(subtest StateSubtest, vmconfig vm.Config, isTestExecutio
 	blockchain.InitDeriveSha(config)
 	block := t.genesis(config).ToBlock(common.Hash{}, nil)
 	memDBManager := database.NewMemoryDBManager()
-	statedb := MakePreState(memDBManager, t.json.Pre)
+	st = MakePreState(memDBManager, t.json.Pre)
 
 	post := t.json.Post[subtest.Fork][subtest.Index]
 	rules := config.Rules(block.Number())
 	msg, err := t.json.Tx.toMessage(post, rules, isTestExecutionSpecState)
 	if err != nil {
-		return nil, err
+		return st, common.Hash{}, err
+	}
+
+	// Try to recover tx with current signer
+	if len(post.TxBytes) != 0 {
+		var ttx types.Transaction
+		txBytes := post.TxBytes
+		if post.TxBytes[0] <= 4 {
+			txBytes = append([]byte{byte(types.EthereumTxTypeEnvelope)}, txBytes...)
+		}
+		err = ttx.UnmarshalBinary(txBytes)
+		if err != nil {
+			return st, common.Hash{}, err
+		}
+		if _, err := types.Sender(types.LatestSigner(config), &ttx); err != nil {
+			return st, common.Hash{}, err
+		}
 	}
+
 	txContext := blockchain.NewEVMTxContext(msg, block.Header(), config)
 	if isTestExecutionSpecState {
 		txContext.GasPrice, err = useEthGasPrice(rules, &t.json)
 		if err != nil {
-			return nil, err
+			return st, common.Hash{}, err
 		}
 	}
 	blockContext := blockchain.NewEVMBlockContext(block.Header(), nil, &t.json.Env.Coinbase)
 	blockContext.GetHash = vmTestBlockHash
-	evm := vm.NewEVM(blockContext, txContext, statedb, config, &vmconfig)
+	if isTestExecutionSpecState {
+		blockContext.GasLimit = t.json.Env.GasLimit
+	}
+	evm := vm.NewEVM(blockContext, txContext, st, config, &vmconfig)
 
 	if isTestExecutionSpecState {
 		useEthOpCodeGas(rules, evm)
 	}
 
-	snapshot := statedb.Snapshot()
+	snapshot := st.Snapshot()
 	result, err := blockchain.ApplyMessage(evm, msg)
 	if err != nil {
-		statedb.RevertToSnapshot(snapshot)
+		st.RevertToSnapshot(snapshot)
```

Comment:
Can we return it here?

Reply:
Sorry, I found this is impossible because this must return non empty root even if err isn't nil.